src/libsyntax/parse/parser.rs

```diff
@@ -1772,7 +1772,12 @@ impl<'a> Parser<'a> {
     /// `Fn::(Args)` (with disambiguator)
     pub fn parse_path(&mut self, style: PathStyle) -> PResult<'a, ast::Path>
     {
-        maybe_whole!(self, NtPath, |x| x);
+        maybe_whole!(self, NtPath, |path| {
+            if style == PathStyle::Mod && path.segments.iter().any(|segment| segment.parameters.is_some()) {
+                self.diagnostic().span_err(path.span, "unexpected generic arguments in path");
+            }
```

Comment:
This won't catch arguments produced by procedural macros, but I guess that's more or less okay.